### DISCUSSION TITLE
Strange behavior when closing overlay for unpoly 3

### DISCUSSION THREAD

**Discussion by spapas** [unpoly/unpoly#473] [Q&A] [answered]
Hello, I'm trying to upgrade my django unpoly demo (https://unpoly-demo.fly.dev/) to use unpoly 3.x.

In the 2.x version for companies section, I had the following snippet to display company detail snippet:

```html
<a
              up-layer='new'
              up-on-dismissed="reloadWithFlashIfEvent('.table', 'Company deleted!', value)"
              up-dismiss-event='company:destroyed'
              href="{% url 'company-detail' company.id %}">{{ company.name }}</a>

```

The `reloadWithFlashIfEvent` is implemented like this:

```javascript
async function reloadWithFlashIfEvent(selector, flash, value) {
    await up.reload(selector, { focus: ':main' })
    if (value instanceof Event || value == ':peel') {
        up.element.affix(document.getElementById('flash-messages'), '.alert.fade.show.alert-danger', { text: flash })
    }
}
```

To dismiss the overlay when the company was deleted I was sending the `company:destroyed` event from the server side.

When I upgraded to unpoly 3.x I noticed the following exception:

```javascript
unpoly.min.js:1 Uncaught (in promise) AbortError: Aborted request to GET /core/tf/: Fragment is being replaced
    at t.setAbortedState (unpoly.min.js:1:75130)
    at t.abort (unpoly.min.js:1:74927)
    at up.Request.Queue.abort (unpoly.min.js:1:82108)
    at Object.a [as abort] (unpoly.min.js:1:136630)
    at Object.abort (unpoly.min.js:1:117280)
    at t.execute (unpoly.min.js:1:16274)
    at t.executePlan (unpoly.min.js:1:6538)
    at t.seekPlan (unpoly.min.js:1:8063)
    at t.execute (unpoly.min.js:1:6441)
    at t.updateContentFromResponse (unpoly.min.js:1:11008)
s
```

which seemed to originate from the `await up.reload(selector, { focus: ':main' })` call since when I commented it out the delete functionality worked fine (it didn't show exception *and* the companies fragment was reloaded)! 

Can you please clarify this behavior?   
   
How should we handle this case? Notice that we need to reload the fragment when the modal closes in order to handle the situation where a company is edited so we need to reload the fragment and get its new name.

Thank you

**Comment by triskweline** [comment]
I didn't have these issues when I upgraded <https://demo.unpoly.com/>. This implementation also reloads fragments after an overlay was closed.

The exception says that a request to `/core/tf` was aborted. What's that URL?

**Reply by spapas** [reply]
This is the url that the delete view redirects to after the delete (and actually it's the list of companies). I guess it gets abandoned because of the reload fragment on the up-on-dismissed event?

Also has the unpoly demo been upgraded to unpoly 3? Last time I checked it was using 2x. If it's been upgraded and works I'll take a look and replicate it.

**Comment by spapas** [comment]
Ok I was able to identify the problem. In my version I had implemented the delete button like this:

```html
      <a class='btn btn-danger' up-layer="new" up-content='
      <h3>Delete company {{ company.name }}</h3>
      Do you want to delete the company? 
      <form up-submit up-layer="root" class="d-inline" method="POST" action="{% url "company-delete" company.id %}">
        {% csrf_token %}
        <input type="submit" value="Yes" class="btn btn-danger mr-3"/>
        <a href="#" class="btn btn-secondary"  up-dismiss>No</a>
      </form>
      '>Delete</a>
```

This opens an html styled popup (instead of the javascript confirmation) with the delete form. Then it will submit the delete form on the *root* layer so response of that view (which are all the companies) will be rendered there and both overlays would be close.

It seems that this behavior conflicts with the on-dismiss=up.reload() event on unpoly 3 (it was working without errors on unpoly 2). I was able to fix it by changing the up-on-dismiss event to the following function:

```javascript

async function reloadWithFlashIfEvent(selector, flash, value) {
    if (value != ':peel') {
        await up.reload(selector, { focus: ':main' })
    }
    if (value instanceof Event || value == ':peel') {
        up.element.affix(document.getElementById('flash-messages'), '.alert.fade.show.alert-danger', { text: flash })
    }
}
```

i.e don't reload if there's a `:peel` event (which is the event that is emitted with the `up-layer='root'`). 

The new version can be seen here: https://unpoly-demo.spapas.net/core/companies/ and works fine.

So the real question is if the error thrown on unpoly 3 (and not thrown on unpoly 2) is expected?

**Reply by triskweline** [reply]
Then it looks like you're loading the companies list twice when the user confirms deletion:

- From the form inside the confirmation, which targets the root layer.
- From the `[up-on-dismissed]` attribute, which calls `up.reload()`.

> So the real question is if the error thrown on unpoly 3 (and not thrown on unpoly 2) is expected?

Since you have two concurrent requests that target the same element (the companies list on the root layer), it's good that Unpoly 3 aborts one of these requests.

As for the custom confirmation dialog, I would suggest a different approach with fewer moving parts.

We put the delete button directly on the company overlay:

```html
<h2>Company 21</h2>

<a href="/companies/21/delete" up-method="post">
```

Now we give the delete button a new `[my-confirm]` attribute indicating that it must be confirmed before it is followed:

```html
<a href="/companies/21/delete" up-method="post" my-confirm="Really delete this company?">
```

We can implement the `[my-confirm]` attribute like this (untested):

```js
up.on('up:link:follow', '[my-confirm]', function(event, link) {
  // Don't follow the link before we confirmed.
  event.preventDefault()

  // Build the HTML content for the confirmation dialog
  let content = `
    <p>${link.getAttribute('my-confirm') || 'Are you sure?'}</p>
    <a up-accept>OK</a>
    <a up-dismiss>Cancel</a>
  `
  
  // When the user confirms, we render with the same options
  // as the link would have rendered.
  let onAccepted = () => up.render(event.renderOptions)

  // Open the confirmation dialog.
  up.layer.open({ content, onAccepted })
})
```

**Comment by spapas** [comment]
Hello @triskweline thank you for answering my question and for the great suggestion!

I tried implementing it your way but I still get an error. Here's what I did:

My delete form (in Django we need to use POST forms for deleting because configuration of the csrf_token with up-method is *not* trivial) is like this:

```html
<form 
        my-confirm="Really??"
        up-submit 
        class="d-inline" method="POST" action="{% url 'company-delete' company.id %}">
        {% csrf_token %}
        <input type="submit" value="Del" class="btn btn-danger mr-3"/>
      </form>
```

and I've got this on my application.js (similar to your suggestion but captures the form submit event):

```javascript

up.on('up:form:submit', '[my-confirm]', function(event, link) {
    // Don't follow the link before we confirmed.
    event.preventDefault()

    // Build the HTML content for the confirmation dialog
    let content = `
      <h3>${link.getAttribute('my-confirm') || 'Are you sure?'}</h3>
      <a class='btn btn-danger' up-accept>Yes</a>
      <a class='btn btn-secondary' up-dismiss>Cancel</a>
    `

    // When the user confirms, we render with the same options
    // as the link would have rendered.
    let onAccepted = () => {
            up.render(event.renderOptions)
        }
        // Open the confirmation dialog.
    up.layer.open({ content, onAccepted })
})

```

It displays the confirmation dialog and works fine when I press cancel, but if I press the "Yes" button I'll get this error:

```javascript

caught (in promise) AbortError: Layer was closed
    at t.abortWhenLayerClosed (unpoly.min.js:1:2445)
    at unpoly.min.js:1:2125
    at up.LayerStack.asCurrent (unpoly.min.js:1:56072)
    at t.asCurrent (unpoly.min.js:1:44526)
    at t.handleLayerChangeRequests (unpoly.min.js:1:2064)
    at t.execute (unpoly.min.js:1:16449)
    at t.executePlan (unpoly.min.js:1:6538)
    at t.seekPlan (unpoly.min.js:1:8063)
    at t.execute (unpoly.min.js:1:6441)
    at t.updateContentFromResponse (unpoly.min.js:1:11008)

```

Any idea what could be the problem? Actually, I tried removing the `up-dismiss-event='company:destroyed'` from the initial link and not it doesn't display that error, however now the overlay won't be dismissed and display non-needed stuff.

Thank you

**Reply by triskweline** [reply, marked as answer]
I believe if you expand that stacktrace you'll find that's just the call `up.render(event.renderOptions)` signaling to you that it could not render its content. The `company:destroyed` event causes the layer to close, so the promise returned `up.render()` will reject with an `AbortError`.

You can ignore that error.

**Reply by spapas** [reply]
Excellent thank you!
